server/reload.go

```diff
@@ -831,53 +829,86 @@ func (s *Server) applyOptions(ctx *reloadContext, opts []option) {
-	oldAccounts := make(map[string]*Account)
-	s.accounts.Range(func(k, v interface{}) bool {
-		acc := v.(*Account)
-		acc.mu.RLock()
-		oldAccounts[acc.Name] = acc
-		acc.mu.RUnlock()
-		s.accounts.Delete(k)
-		return true
-	})
-	s.gacc = nil
-	s.configureAccounts()
-
-	s.configureAuthorization()
-
 	// This map will contain the names of accounts that have their streams
 	// import configuration changed.
 	awcsti := make(map[string]struct{})
 
-	s.accounts.Range(func(k, v interface{}) bool {
-		newAcc := v.(*Account)
-		if acc, ok := oldAccounts[newAcc.Name]; ok {
-			// If account exist in latest config, "transfer" the account's
-			// sublist and client map to the new account.
+	s.mu.Lock()
+
+	// This can not be changed for now so ok to check server's trustedKeys
+	if s.trustedKeys == nil {
+		// We need to drain the old accounts here since we have something
+		// new configured. We do not want s.accounts to change since that would
+		// mean adding a lock to lookupAccount which is what we are trying to
+		// optimize for with the change from a map to a sync.Map.
+		oldAccounts := make(map[string]*Account)
+		s.accounts.Range(func(k, v interface{}) bool {
+			acc := v.(*Account)
 			acc.mu.RLock()
-			if len(acc.clients) > 0 {
-				newAcc.clients = make(map[*client]*client, len(acc.clients))
-				for _, c := range acc.clients {
-					newAcc.clients[c] = c
+			oldAccounts[acc.Name] = acc
+			acc.mu.RUnlock()
+			s.accounts.Delete(k)
+			return true
+		})
+		s.gacc = nil
+		s.configureAccounts()
+		s.configureAuthorization()
+
+		s.accounts.Range(func(k, v interface{}) bool {
+			newAcc := v.(*Account)
+			if acc, ok := oldAccounts[newAcc.Name]; ok {
+				// If account exist in latest config, "transfer" the account's
+				// sublist and client map to the new account.
+				acc.mu.RLock()
+				if len(acc.clients) > 0 {
+					newAcc.clients = make(map[*client]*client, len(acc.clients))
+					for _, c := range acc.clients {
+						newAcc.clients[c] = c
+					}
+				}
+				newAcc.sl = acc.sl
+				newAcc.rm = acc.rm
+				acc.mu.RUnlock()
+
+				// Check if current and new config of this account are same
+				// in term of stream imports.
+				if !acc.checkStreamImportsEqual(newAcc) {
+					awcsti[newAcc.Name] = struct{}{}
 				}
 			}
-			newAcc.sl = acc.sl
-			acc.mu.RUnlock()
+			return true
+		})
+	} else if s.opts.AccountResolver != nil {
+		s.configureResolver()
+		if _, ok := s.accResolver.(*MemAccResolver); ok {
+			// With a memory resolver we want to do something similar to configured accounts.
+			// We will walk the accounts and delete them if they are no longer present via fetch.
+			// If they are present we will force a claim update to process changes.
+			s.accounts.Range(func(k, v interface{}) bool {
+				acc := v.(*Account)
+				// Skip global account.
+				if acc == s.gacc {
+					return true
+				}
+				acc.mu.RLock()
+				accName := acc.Name
+				acc.mu.RUnlock()
+				accClaims, claimJWT, _ := s.fetchAccountClaims(accName)
```

Comment:
This will release server lock. Does that introduce a race? See what we did in fetchAccount() because of that.

Reply:
We are checking accounts but yes the fetch will release lock.

Reply:
Yes, ok, that's different than the issue I was thinking about. In referenced issue, the problem was concurrent lookup of an account, in this case, we know we have the account, just fetch the claim.